### PR TITLE
GroupBy(multiple groupers)

### LINKED ISSUE #1056: groupby_bins along two dims simultaneously
I probably missed it, but what is the way to apply groupby (or rather groupby_bins) in order to achieve the following in xarray?

``` python
da = xr.DataArray(np.arange(16).reshape((4, 4)))
da
<xarray.DataArray (dim_0: 4, dim_1: 4)>
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])
Coordinates:
  * dim_0    (dim_0) int64 0 1 2 3
  * dim_1    (dim_1) int64 0 1 2 3

# should be aggregated to (in case of summing) to obtain
dagg
<xarray.DataArray (dim_0: 2, dim_1: 2)>
array([[10, 18],
       [42, 50]])
Coordinates:
  * dim_1    (dim_1) int64 0 2
  * dim_0    (dim_0) int64 0 2
```

### LINKED ISSUE #9332: Enable multi-coord grouping from xarray
I'm working through some around non-trivial `groupby` from colleagues. In particular, grouping by multiple coordinates seems much harder in xarray than pandas. Flox actually does this really nicely, as per the comment from @dcherian:

> As an aside, the API isn't great but this works in flox (I think)

```python
import flox.xarray

flox.xarray.xarray_reduce(da, "labels1", "labels2", func="mean")
```
image

_Originally posted by @dcherian in https://github.com/pydata/xarray/issues/9278#issuecomment-2250993016_

Would be great if `da.groupby(["labels1", "labels2"]).mean()` worked too — is that just a simple translation to flox from the xarray code? (I can probably do it if so). Or is there something more complex going on?

### PR DESCRIPTION
**Pull request by dcherian**
Builds on #9389 

- [x] Closes #924, 
- [x] Closes #1056, 
- [x] Closes #9332, 
- [x] xref #324
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`
- [x] add narrative docs
- [ ] Update docstrings

-----

This PR takes an approach similar to that in #924. 

The GroupBy(multiple array) problem can be reduced to the GroupBy(single nD variable) problem:
1. First convert each `by` array to integer codes ("factorize"). I do this first because factorization is expensive, and so it's cheaper to do it before any possible broadcasting.
2.  Broadcast DataArrays of integer codes against each other.
3. Then `np.ravel_multi_index` them to construct a single nD array of integer codes.
4. At this point, we treat it as a GroupBy(single nD array) problem i.e. reshape down to 1D problem. This isn't great for dask/cubed but it is simple and matches what we do today. That means we can apply all the operations we do with GroupBy today.
    - Importantly UDFs receive "stacked arrays" as they do today when grouping by a nD array. This isn't the most elegant but it preserves dtype. For example, you can't index out a 2D array with `group==1` without inserting a missing value if `group` is:
    ```
    [[1, 1, 1],
     [2, 1, 1],
     [2, 2, 1]]
     ```
5. Once operations are complete, we assign a new MultiIndex and then unstack back to restore the needed dimensions.
    - I'm not sure this is the best at the moment - what if one of the arrays we group by is a multiIndexed array?

TODO:
- [x] lots more tests!

### CODE REVIEW

**Comment by dcherian** [comment]
@max-sixty I could use some help with adding tests if you have time :)

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -317,33 +316,7 @@ def size(self) -> int:
 
     def __len__(self) -> int:
         """Number of groups."""
-        return len(self.full_index)
-
-    @property
-    def dims(self):
-        return self.group1d.dims
-
-    def factorize(self) -> None:
```
Comment (dcherian) [inline xarray/core/groupby.py:326]: Moved to `EncodedGroups.__init__`

File: xarray/groupers.py [hunk 2]
```diff
@@ -57,18 +67,49 @@ class EncodedGroups:
...
+
+        assert isinstance(codes, DataArray)
+        if codes.name is None:
             raise ValueError("Please set a name on the array you are grouping by.")
-        assert isinstance(self.full_index, pd.Index)
-        assert (
-            isinstance(self.unique_coord, Variable | _DummyGroup)
-            or self.unique_coord is None
-        )
+        self.codes = codes
```
Comment (dcherian) [inline xarray/groupers.py:87]: moved from `ResolvedGrouper.factorize`

File: xarray/core/groupby.py [hunk 3]
```diff
@@ -410,6 +382,58 @@ def _resolve_group(
     return newgroup
 
 
+@dataclass
```
Comment (dcherian) [inline xarray/core/groupby.py:385]: This is the main new logic.

File: xarray/core/groupby.py [hunk 4]
```diff
@@ -479,26 +509,47 @@ def __init__(
...
 
-        (grouper,) = self.groupers
-        self._original_group = grouper.group
+        if len(groupers) == 1:
+            (grouper,) = groupers
+            self.encoded = grouper.encoded
+        else:
+            for grouper in groupers:
+                if grouper.group.ndim > 1:
+                    raise NotImplementedError(
```
Comment (dcherian) [inline xarray/core/groupby.py:522]: This just needs some tests to enable.

**Comment by max-sixty** [comment]
I had a look at where I add something — it already seems extremely good? `gb - gb.mean()` would be cool but this already seems like a big step forwards.

I missed previous discussions / can decide this later — but do we want to have `da.groupby(['foo', 'bar'])` as sugar for `da.groupby(foo=UniqueGrouper(), bar=UniqueGrouper())`?

**Comment by dcherian** [comment]
Yes it is surprisingly complete. 

A section in the narrative docs would be nice to add.

**Comment by dcherian** [comment]
Yes it is surprisingly complete. 

A section in the narrative docs would be nice to add.

File: xarray/core/groupby.py (hunk 4 above)
Comment (dcherian) [reply xarray/core/groupby.py:522]: @max-sixty a test for this path would be good. I can fix it up if it doesn't work

**Comment by dcherian** [comment]
> missed previous discussions / can decide this later — but do we want to have da.groupby(['foo', 'bar']) as sugar for da.groupby(foo=UniqueGrouper(), bar=UniqueGrouper())?

I agree. It's a bit too much typing at the moment

**Comment by dcherian** [comment]
We should test grouping by a multi-index array and some other array but otherwise this should be good to go.

File: doc/user-guide/groupby.rst [hunk 5]
```diff
@@ -81,9 +81,11 @@ You can index out a particular group:
 
     ds.groupby("letters")["b"]
 
-Just like in pandas, creating a GroupBy object is cheap: it does not actually
+Just like in pandas, creating a ``GroupBy`` object is cheap: it does not actually
 split the data until you access particular values.
 
+To group by multiple variables, see the section on `Grouper Objects`_
+
```
Comment (dcherian) [inline doc/user-guide/groupby.rst:88]:
```suggestion
To group by multiple variables, see :ref:`this section <groupby.multiple>`.

```

**Review by max-sixty — APPROVED** [review]
This is huge! Thank you very much @dcherian !

**Comment by max-sixty** [comment]
Huge success! Thank you @dcherian !

**Comment by shoyer** [comment]
Wow, amazing work @dcherian! I am so excited to see this finally land.

**Pull request merged**